File: .credentials/tests/test_get_credentials.R
# Run:     Rscript .credentials/tests/test_get_credentials.R
# Inputs:  Temporary credential files created during the test
# Outputs: Test status; no project files are modified

script_arg <- grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)
script_file <- sub("^--file=", "", script_arg[[1]])
source(file.path(dirname(normalizePath(script_file)), "..", "get-credentials.R"))

secret_file <- tempfile()
.mounted_credentials_file <- secret_file
old_value <- Sys.getenv("TEST_TOKEN", unset = NA_character_)
on.exit({
  if (is.na(old_value)) Sys.unsetenv("TEST_TOKEN") else Sys.setenv(TEST_TOKEN = old_value)
  unlink(secret_file)
}, add = TRUE)

writeLines("TEST_TOKEN=mounted", secret_file)
Sys.setenv(TEST_TOKEN = "environment")
stopifnot(identical(get_credential("TEST_TOKEN", "test-token"), "environment"))

Sys.unsetenv("TEST_TOKEN")
writeLines(c("# comment", "TEST_TOKEN=value=with=equals", "MALFORMED"), secret_file)
stopifnot(identical(get_credential("TEST_TOKEN", "test-token"), "value=with=equals"))

writeLines("TEST_TOKEN=mounted", secret_file)
stopifnot(identical(get_credential("TEST_TOKEN", "test-token"), "mounted"))
writeLines(character(), secret_file)
old_store_reader <- .read_from_store
.read_from_store <- function(target) "host-value"
stopifnot(identical(get_credential("TEST_TOKEN", "test-token"), "host-value"))
.read_from_store <- old_store_reader

message("All R credential tests passed.")

File: .credentials/tests/test_get_credentials.py
#!/usr/bin/env python3
# Run:     python3 .credentials/tests/test_get_credentials.py
# Inputs:  Temporary credential files created by each test
# Outputs: unittest results; no project files are modified

import importlib.machinery
import importlib.util
import os
import tempfile
import unittest
from pathlib import Path
from unittest import mock


MODULE_PATH = Path(__file__).resolve().parents[1] / "get-credentials.py"


def load_module():
    spec = importlib.util.spec_from_loader(
        "get_credentials_under_test",
        importlib.machinery.SourceFileLoader(
            "get_credentials_under_test",
            str(MODULE_PATH),
        ),
    )
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


class CredentialResolutionTests(unittest.TestCase):
    def setUp(self):
        self.module = load_module()
        self.temp_dir = tempfile.TemporaryDirectory()
        self.addCleanup(self.temp_dir.cleanup)
        self.secret_file = Path(self.temp_dir.name) / "credentials.env"
        self.module._MOUNTED_CREDENTIALS_FILE = self.secret_file

    def test_environment_overrides_mounted_file(self):
        self.secret_file.write_text("TEST_TOKEN=mounted\n", encoding="utf-8")
        with mock.patch.dict(os.environ, {"TEST_TOKEN": "environment"}):
            value = self.module.get_credential("TEST_TOKEN", "test-token")
        self.assertEqual(value, "environment")

    def test_reads_values_from_mounted_file(self):
        self.secret_file.write_text(
            "# comment\n"
            "TEST_TOKEN=value=with=equals\n"
            "MALFORMED\n",
            encoding="utf-8",
        )
        with mock.patch.dict(os.environ, {}, clear=True):
            value = self.module.get_credential("TEST_TOKEN", "test-token")
        self.assertEqual(value, "value=with=equals")

    def test_reads_file_afresh_so_updates_are_visible(self):
        self.secret_file.write_text("TEST_TOKEN=mounted\n", encoding="utf-8")
        with mock.patch.dict(os.environ, {}, clear=True):
            self.assertEqual(
                self.module.get_credential("TEST_TOKEN", "test-token"),
                "mounted",
            )
            self.secret_file.write_text("", encoding="utf-8")
            with mock.patch.object(
                self.module,
                "_read_from_store",
                return_value="",
            ):
                self.assertEqual(
                    self.module.get_credential("TEST_TOKEN", "test-token"),
                    "",
                )

    def test_falls_back_to_host_store_when_file_is_absent(self):
        with (
            mock.patch.dict(os.environ, {}, clear=True),
            mock.patch.object(
                self.module,
                "_read_from_store",
                return_value="host-value",
            ),
        ):
            value = self.module.get_credential("TEST_TOKEN", "test-token")
        self.assertEqual(value, "host-value")


if __name__ == "__main__":
    unittest.main()
